[CanonicalizeOSSALifetime] Extend lexical lifetimes to unreachables.

nate-chandler opened the pull request:
When canonicalizing the lifetime of a lexical value, deinit barriers are respected. This is done by walking backwards from lifetime ends and adding encountered deinit barriers to liveness.

Only destroy lifetime ends were walked back from under the assumption that lifetimes would be complete. Without complete OSSA lifetimes, however, it's necessary to also necessary to consider lifetimes that end with unreachables. Unfortunately, we can't simply walk back from those unreachables because there may be instructions which are secretly users of the value being canonicalized (e.g. destroys of `partial_apply`s to which a `begin_borrow` of the value was passed). Such uses don't appear in the use list because lifetime canonicalization expects complete lifetimes and only visits lifetime ends of `begin_borrow`s.

Here, instead, the instructions before the relevant unreachables are added to liveness. In order to determine which unreachables are relevant, it's necessary to have a liveness that includes the original destroys. So a copy of liveness is created and those destroys are added to it.

rdar://115468707

nate-chandler committed:
[PrunedLiveness] Gardening: Deleted TODO.

It was already addressed.

nate-chandler committed:
[PrunedLiveness] Added copy constructor.

It's useful to be able to form a second copy of liveness which is
differently pruned by having more instructions added to it, e.g..

nate-chandler committed:
[CanOSSALifetime] Respect lexical lifetimes flag.

Don't respect deinit barriers when canonicalizing if lexical lifetimes
are disabled.

nate-chandler committed:
[OSSALifeComp] NFC: Extracted visit unreachable.

Extracted the new visitUnreachableLifetimeEnds static member of
OSSALifetimeCompletion from the preexisting
endLifetimeAtUnreachableBlocks which now calls through the former.

nate-chandler committed:
[CanOSSALife] Lexical lifetimes go to unreachables

When canonicalizing the lifetime of a lexical value, deinit barriers are
respected. This is done by walking backwards from lifetime ends and
adding encountered deinit barriers to liveness.

Only destroy lifetime ends were walked back from under the assumption
that lifetimes would be complete. Without complete OSSA lifetimes,
however, it's necessary to also necessary to consider lifetimes that end
with unreachables. Unfortunately, we can't simply walk back from those
unreachables because there may be instructions which are secretly users
of the value being canonicalized (e.g. destroys of `partial_apply`s to
which a `begin_borrow` of the value was passed). Such uses don't appear
in the use list because lifetime canonicalization expects complete
lifetimes and only visits lifetime ends of `begin_borrow`s.

Here, instead, the instructions before the relevant unreachables are
added to liveness. In order to determine which unreachables are
relevant, it's necessary to have a liveness that includes the original
destroys. So a copy of liveness is created and those destroys are added
to it.

rdar://115468707

nate-chandler commented:
@swift-ci please test

atrick commented on lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:
```diff
@@ -249,6 +250,33 @@ void CanonicalizeOSSALifetime::extendLivenessToDeinitBarriers() {
+      getCurrentDef(), completeLiveness, [&](auto *unreachable) {
+        recordConsumingUser(unreachable);
+        if (auto *previous = unreachable->getPreviousInstruction()) {
+          if (liveness->isInterestingUser(previous) ==
```
Let's avoid this bug-prone check in a separate refactoring. Maybe PrunedLiveness needs an extendToNonUse API

nate-chandler replied:
https://github.com/apple/swift/pull/68756

atrick reviewed (approved):
Looks good!

nate-chandler committed:
[SILCombine] Keep lexical destroys post cond_fails

Such destroys are part of the lifetime of lexical values inserted to
prevent shortening lifetimes over deinit baarriers by
CanonicalizeOSSALifetime.  After OSSALifetimeCompletion is enabled, more
instructions will need to be preserved.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test windows platform
